### PR TITLE
Oliver Pérez - UI Design Patterns Challenge

### PR DESCRIPTION
**Pull request by oliver-perez**

### CODE REVIEW

File: MVC Module/Source/Feed Presentation/en.lproj/Feed.strings [hunk 1]
```diff
@@ -1,3 +1,3 @@
 
-"FEED_VIEW_TITLE" = "My Feed";
-"FEED_VIEW_CONNECTION_ERROR" = "Couldn't connect to server";
+"FEED_VIEW_TITLE" = "Mi Feed";
+"FEED_VIEW_CONNECTION_ERROR" = "No se pudo conectar al servidor";
```
Comment (caiozullo) [inline MVC Module/Source/Feed Presentation/en.lproj/Feed.strings:3]: This file should contain the English translations - so it shouldn't be modified. Revert the changes.

File: MVP Module/Source/Feed Presentation/FeedPresenter.swift [hunk 2]
```diff
@@ -11,6 +11,8 @@ protocol FeedLoadingView {
 
 protocol FeedView {
 	func display(_ viewModel: FeedViewModel)
+	func showError(message: String)
+	func hideError()
```
Comment (caiozullo) [inline MVP Module/Source/Feed Presentation/FeedPresenter.swift:15]: Ideally, the error presentation should be in a new protocol, such as a `FeedErrorView` because we may display different kinds of errors (not just feed loading errors).

Moreover, we should avoid changing existing protocols when adding new features because it affect existing code. For example, the Adapter had to be changed in this PR, even though it has nothing to do with errors. Ideally, *the adapter should have been unchanged.*

File: MVP Module/Source/Feed UI/Composers/FeedViewAdapter.swift [hunk 3]
```diff
@@ -26,4 +26,12 @@ final class FeedViewAdapter: FeedView {
...
 		})
 	}
+	
+	func hideError() {
+		controller?.hideError()
+	}
+	
+	func showError(message: String) {
+		controller?.showError(message: message)
+	}
```
Comment (caiozullo) [inline MVP Module/Source/Feed UI/Composers/FeedViewAdapter.swift:36]: Ideally, the adapter shouldn't have been changed. See my other comment.

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift [hunk 4]
```diff
@@ -20,12 +20,15 @@ final class FeedViewModel {
 	
 	var onLoadingStateChange: Observer<Bool>?
 	var onFeedLoad: Observer<[FeedImage]>?
+	var onLoadError: Observer<String>?
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/FeedViewModel.swift:23]: `onLoadError: Observer<String> ` could be `onErrorStateChange: Observer<String?>`, so the ViewModel decides which message to show and when to show/hide errors (instead of the view controller). It's often easier to manage state when all changes are centralized.

For example, in the `loadFeed()` method, the ViewModel can call  `onErrorStateChange(.none)` to hide the error before loading the feed. And when there's an error: `onErrorStateChange(Localized.Feed.errorMessage)`.

File: MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 5]
```diff
@@ -27,11 +30,16 @@ public final class FeedViewController: UITableViewController, UITableViewDataSou
 		title = viewModel?.title
 		viewModel?.onLoadingStateChange = { [weak self] isLoading in
 			if isLoading {
+				self?.errorView?.hideMessage()
```
Comment (caiozullo) [inline MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift:33]: The ViewModel should decide when to hide/show errors. See my other comment.

File: MVVM Module/Tests/Feed UI/Helpers/FeedViewController+TestHelpers.swift [hunk 6]
```diff
@@ -47,6 +47,17 @@ extension FeedViewController {
 	func numberOfRenderedFeedImageViews() -> Int {
 		return tableView.numberOfRows(inSection: feedImagesSection)
 	}
+	private var errorView: ErrorView? {
```
Comment (caiozullo) [inline MVVM Module/Tests/Feed UI/Helpers/FeedViewController+TestHelpers.swift:50]: Missing line break:

```suggestion
	}
	
	private var errorView: ErrorView? {
```

**Review by caiozullo — COMMENTED** [review]
Hi @oliver-perez, you almost passed the challenge!

Please, check my review comments and let me know once you're done so I can review it again. Thank you 🙂

**Comment by oliver-perez** [comment]
Hi @caiozullo, I've just attended to your comments. 

Thank you for the feedback 😀

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift [hunk 7]
```diff
@@ -20,12 +20,17 @@ final class FeedViewModel {
 	
 	var onLoadingStateChange: Observer<Bool>?
 	var onFeedLoad: Observer<[FeedImage]>?
+	var onErrorStateChange: Observer<String?> = { _ in }
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/FeedViewModel.swift:23]: The `onErrorStateChange` property doesn't need a default value. Just like the other closure properties in this class, it can be optional:

```suggestion
	var onErrorStateChange: Observer<String?>?
```

**Review by caiozullo — COMMENTED** [review]
Hi @oliver-perez,

Congratulations, you've passed the challenge 🥇

- All tests are passing ✅
- Your code is clean and handles edge cases ✅
- The UI and animations look and behave as expected ✅
- Light and Dark mode support ✅
- Added an extra language ✅

I left a review comment. Feel free to push changes.

Great job!
